SCClassLibrary/Common/Control/ReadableNodeIDAllocator.sc

```diff
@@ -0,0 +1,72 @@
+/* make nodeIDs more humanly readable:
+10 ** 8 ids for 1 user, 10 ** 7 for < 21, 10 ** 6 for < 210
+
+for example, with 20 Clients, clientIDs would be
+clientID  defaultGroup  tempIDs
+ 1         10000001      120001000, 10001001, etc
+ 2         10000001      120001000, 10001001, etc
+12        120000001      120001000, 10001001, etc
+
+In multi-client settings, this makes s.plotTree display
+or s.queryAllNodes posts much easier to understand.
+*/
+
+ReadableNodeIDAllocator {
+
+	var <clientID, <lowestTempID, <numClients;
+	var <numIDs, <idOffset, <maxPermID;
+	var tempCount = -1, permCount = 1, permFreed;
+
+	*new { arg clientID = 0, lowestTempID = 1000, numClients = 32;
+		^super.newCopyArgs(clientID, lowestTempID, numClients).reset
+	}
+
+	reset {
+		var maxNumPerUser = (2 ** 31 / numClients).trunc;
+		// go to next lower power of 10:
+		numIDs = (10 ** maxNumPerUser.log10.trunc).asInteger;
+
+		idOffset = numIDs * clientID;
+
+		permFreed = IdentitySet.new;
+		maxPermID = idOffset + lowestTempID - 1;
+		tempCount = -1;
+		permCount = 1;
+	}
+
+	alloc {
+		tempCount = tempCount + 1;
+		// wraps after handing out (numIDs - lowestTempID) ids
+		^(lowestTempID + tempCount).wrap(lowestTempID, numIDs) + idOffset;
+	}
+
+	isPerm { |num|
+		// 0 and 1 are also permanent
```

Comment:
i'm not sure i understand this comment, or at least it doesn't help me as a first-time reader

Reply:
ok, I'll try again: 
the allocator creates a sequence of nodeIDs starting within the range of 
lowestTempID (1000) to numIDs (a power of 10, say 10000000).
when the maximum nodeID is reached, we wrap back in the range, starting with to lowestTempID, 
assuming all the really low nodeIDs are long gone by then (after say a week of running an installation with massive granular synthesis). 
the entire range is shifted by idOffset, which creates the prefix of say 12 (clientID) * 1000000 (numIDs) = 120001000, 120001001 etc.

is that clearer? sorry if I am too deep into this to explain it well :-)
would it be better to put the longer story into the help file, and get rid of the comment here?

Reply:
the comment we're talking about says `0 and 1 are also permanent`... i already understand what you just explained in your last comment, because that info is already at the top of this file. what you're saying doesn't seem to be clearly related to the comment i'm asking about, though - you didn't mention anything about 0, 1, or permanent node allocation.

whatever this comment is supposed to mean, it is definitely not clear enough. i think what i said above is correct - that by using `inclusivelyBetween`, the code is sure to catch permanent node IDs that are offset at 0 and 1 from the `idOffset`. is that right?

thanks for your patience.

Reply:
OK, now I got that you meant the comment in isPerm, sorry for my slowness!
clarified the comment in 0beaddd, hope this is fine now.